sklearn_numba_dpex/kmeans/kernels.py

```diff
@@ -0,0 +1,727 @@
+    zero = dtype(0)
+
+    @dpex.kernel
+    def initialize_to_zeros(x):
+        sample_idx = dpex.get_global_id(0)
+
+        if sample_idx >= n_samples:
+            return
+
+        x[sample_idx] = zero
+
+    return initialize_to_zeros[global_size, work_group_size]
+
+
+@lru_cache
+def make_initialize_to_zeros_2dim_kernel(n_samples, n_features, work_group_size, dtype):
+
+    n_items = n_samples * n_features
+    global_size = math.ceil(n_items / work_group_size) * work_group_size
+    zero = dtype(0.0)
+
+    @dpex.kernel
+    def initialize_to_zeros(x):
+        item_idx = dpex.get_global_id(0)
+
+        if item_idx >= n_items:
+            return
+
+        i = item_idx // n_samples
+        j = item_idx % n_samples
+        x[i, j] = zero
+
+    return initialize_to_zeros[global_size, work_group_size]
+
+
+@lru_cache
+def make_initialize_to_zeros_2dim_kernel(n_samples, n_features, work_group_size, dtype):
+
+    n_items = n_samples * n_features
+    global_size = math.ceil(n_items / work_group_size) * work_group_size
+    zero = dtype(0.0)
+
+    @dpex.kernel
+    def initialize_to_zeros(x):
+        item_idx = dpex.get_global_id(0)
+
+        if item_idx >= n_items:
+            return
+
+        i = item_idx // n_samples
+        j = item_idx % n_samples
+        x[i, j] = zero
+
+    return initialize_to_zeros[global_size, work_group_size]
+
+
+@lru_cache
+def make_initialize_to_zeros_3dim_kernel(dim0, dim1, dim2, work_group_size, dtype):
+
+    n_items = dim0 * dim1 * dim2
+    stride0 = dim1 * dim2
+    global_size = math.ceil(n_items / work_group_size) * work_group_size
+    zero = dtype(0.0)
+
+    @dpex.kernel
+    def initialize_to_zeros(x):
+        item_idx = dpex.get_global_id(0)
+
+        if item_idx >= n_items:
+            return
+
+        i = item_idx // stride0
+        stride0_idx = item_idx % stride0
+        j = stride0_idx // dim2
+        k = stride0_idx % dim2
+        x[i, j, k] = zero
+
+    return initialize_to_zeros[global_size, work_group_size]
+
+
+@lru_cache
+def make_copyto_kernel(n_samples, n_features, work_group_size):
+    global_size = math.ceil(n_samples / work_group_size) * work_group_size
+
+    @dpex.kernel
+    def copyto_kernel(X, Y):
+        sample_idx = dpex.get_global_id(0)
+
+        if sample_idx >= n_samples:
+            return
+
+        for feature_idx in range(n_features):
+            Y[feature_idx, sample_idx] = X[feature_idx, sample_idx]
+
+    return copyto_kernel[global_size, work_group_size]
+
+
+@lru_cache
+def make_broadcast_division_kernel(n_samples, n_features, work_group_size):
+    global_size = math.ceil(n_samples / work_group_size) * work_group_size
+
+    @dpex.kernel
+    def broadcast_division(X, v):
+        sample_idx = dpex.get_global_id(0)
+
+        if sample_idx >= n_samples:
+            return
+
+        divisor = v[sample_idx]
+
+        for feature_idx in range(n_features):
+            X[feature_idx, sample_idx] = X[feature_idx, sample_idx] / divisor
```

Comment:
I'm enclined to say that it doesn't matter (providing nan behaves in the kernel as it does in numpy regarding comparisons), and I suspect it can't happen if the initialization does not start with empty clusters (e.g with random initialization which choose data points as starting clusters), there might be a theoretical result for this ?

If the initialization creates a cluster with 0 points, I guess all the subsequent iterations would result as all the values of the corresponding centroid being set to nan, without impacting the result otherwise.

But it's a good point, I can add a unit test to see what happens.